src/analysis/reaching-definitions-transfer-function.h

```diff
@@ -0,0 +1,176 @@
+#ifndef wasm_analysis_reaching_definitions_transfer_function_h
+#define wasm_analysis_reaching_definitions_transfer_function_h
+
+#include "ir/local-graph.h"
+#include "lattice.h"
+#include "monotone-analyzer.h"
+#include "wasm-traversal.h"
+
+namespace wasm::analysis {
+
+class ReachingDefinitionsTransferFunction
+  : public Visitor<ReachingDefinitionsTransferFunction> {
+  // In addition to LocalSet expressions modifying a local index's value, we
+  // must also account for the fact that the local index's default value at
+  // initialization is used, or that it already has a value at the beginning.
+  // Therefore, when creating the powerset lattice from the list of all
+  // LocalSetses, we also append a list of nullptrs  numLocals long, one for
+  // each local index, at the end of the list of all LocalSetses. These nullptrs
+  // represent the use of a default/param value.
+
+  FinitePowersetLattice<LocalSet*>::Element* currState = nullptr;
+  FinitePowersetLattice<LocalSet*>& lattice;
+
+  // Number of locals in a function.
+  size_t numLocals;
+
+  // Maps a local index to a list of LocalSets that modify the index's value.
+  std::unordered_map<Index, SmallVector<LocalSet*, 2>> indexSetses;
+
+  // LocalGraph members we need to update.
+  LocalGraph::GetSetses* getSetses = nullptr;
+
+  // We actually don't update locations right now since the CFG we are working
+  // with doesn't contain the correct Expression**s, but this is left in for
+  // future improvements.
+  LocalGraph::Locations* locations = nullptr;
+
+  // Signals to the visit functions whether we are solving the analysis, or
+  // collecting results from the solved analysis.
+  bool collectingResults = false;
+
+public:
+  ReachingDefinitionsTransferFunction(FinitePowersetLattice<LocalSet*>& lattice,
+                                      size_t numLocals)
+    : lattice(lattice), numLocals(numLocals) {
+
+    // Populate indexSetses.
+    for (auto it = lattice.membersBegin();
+         it != lattice.membersEnd() - numLocals;
+         ++it) {
+      indexSetses[(*it)->index].push_back(*it);
+    }
+  }
+
+  // Sets the LocalGraph members we want to update, and signals that result
+  // collection is about to begin.
+  void beginResultCollection(LocalGraph::GetSetses* getSetses,
+                             LocalGraph::Locations* locations) {
+    this->getSetses = getSetses;
+    this->locations = locations;
+    collectingResults = true;
+  }
```

Comment:
This isn't used in this PR because it is used in `LocalGraph` to collect results into the `LocalGraph` members. I will change the test case to collect results in the same manner as `LocalGraph` instead of via printing.